src/Util/DynamicLinker.hs

```diff
@@ -69,13 +74,65 @@ libFileName dirs lib = do let names = [lib, lib ++ "." ++ hostDynamicLibExt]
                           return $ maybe (lib ++ "." ++ hostDynamicLibExt) id found
 
 #ifndef mingw32_HOST_OS
+-- Load a dynamic library on POSIX systems.
+-- In the simple case, we just find the appropriate filename and call dlopen().
+-- In the complicated case our "foo.so" isn't actually a library. Some of the
+-- .so files on modern Linux systems are linker scripts instead. dlopen()
+-- doesn't know anything about those. We need to look inside the script for the
+-- actual library path and load that. This is a horrible hack, the correct
+-- method would be to actually parse the scripts and execute them. The approach
```

Comment:
The “correct” method would not in fact be to parse the scripts, since the scripts are for the *static* linker. (When the static linker encounters such a script, in the common case it will link whatever it’s working on with the shared object named in the script directly, possibly additionally linking in static libraries when necessary.) There is actually a C header file with `#define`s in it naming the actual shared objects of the installed glibc; these are what a `dlopen()`-using program is supposed to use according to glibc devs.